tikv/gc.go

```diff
@@ -94,72 +95,124 @@ func (s *KVStore) resolveLocks(ctx context.Context, safePoint uint64, concurrenc
-const gcScanLockLimit = txnlock.ResolvedCacheSize / 2
+type BaseLockResolver struct {
+	identifier string
+	store      Storage
+}
+
+func NewBaseLockResolver(identifier string, store Storage) *BaseLockResolver {
+	return &BaseLockResolver{
+		identifier: identifier,
+		store:      store,
+	}
+}
+
+func (l *BaseLockResolver) Identifier() string {
+	return l.identifier
+}
+
+func (l *BaseLockResolver) ResolveLocks(bo *Backoffer, locks []*txnlock.Lock, loc *locate.KeyLocation) (*locate.KeyLocation, error) {
+	return batchResolveLocksInARegion(bo, l.GetStore(), locks, loc)
+}
+
+func (l *BaseLockResolver) ScanLocks(bo *Backoffer, key []byte, maxVersion uint64, scanLimit uint32) ([]*txnlock.Lock, *locate.KeyLocation, error) {
+	return scanLocksInRegionWithStartKey(bo, l.GetStore(), key, maxVersion, scanLimit)
+}
 
-func (s *KVStore) resolveLocksForRange(ctx context.Context, safePoint uint64, startKey []byte, endKey []byte) (rangetask.TaskStat, error) {
+func (l *BaseLockResolver) GetStore() Storage {
+	return l.store
+}
+
+// GCLockResolver is used for GCWorker and log backup advancer to resolve locks.
+type GCLockResolver interface {
+	Identifier() string
+	// ResolveLocks tries to resolve expired locks.
+	// 1. For GCWorker it will scan locks for all regions before *safepoint*,
+	// and force remove locks. rollback the txn, no matter the lock is expired of not.
+	// 2. For log backup advancer, it will scan all locks for a small range.
+	// and it will check status of the txn. resolve the locks if txn is expired, Or do nothing.
+	ResolveLocks(*Backoffer, []*txnlock.Lock, *locate.KeyLocation) (*locate.KeyLocation, error)
+
+	// ScanLocks return locks and location with given start key
+	ScanLocks(*Backoffer, []byte, uint64, uint32) ([]*txnlock.Lock, *locate.KeyLocation, error)
+
+	// GetStore is used to get store to GetRegionCache and SendReq for this lock resolver.
+	GetStore() Storage
+}
+
+// We don't want gc to sweep out the cached info belong to other processes, like coprocessor.
+const GCScanLockLimit = txnlock.ResolvedCacheSize / 2
+
+func ResolveLocksForRange(
+	ctx context.Context,
+	resolver GCLockResolver,
+	maxVersion uint64,
+	startKey []byte,
+	endKey []byte,
+	createBackoffFn func(context.Context) *Backoffer,
+	scanLimit uint32,
+) (rangetask.TaskStat, error) {
 	// for scan lock request, we must return all locks even if they are generated
 	// by the same transaction. because gc worker need to make sure all locks have been
 	// cleaned.
-
 	var stat rangetask.TaskStat
 	key := startKey
-	bo := NewGcResolveLockMaxBackoffer(ctx)
 	for {
 		select {
 		case <-ctx.Done():
 			return stat, errors.New("[gc worker] gc job canceled")
 		default:
 		}
-
-		locks, loc, err := s.scanLocksInRegionWithStartKey(bo, key, safePoint, gcScanLockLimit)
+		// create new backoffer for every scan and resolve locks
+		bo := createBackoffFn(ctx)
+		locks, loc, err := resolver.ScanLocks(bo, key, maxVersion, scanLimit)
```

Comment:
TiDB has this test code: 
```go
		if w.testingKnobs.scanLocks != nil {
			locks = append(locks, w.testingKnobs.scanLocks(key, loc.Region.GetID(), tryResolveLocksTS)...)
		}
```
This is used to inject some simulated locks for tests. Did you notice it, and do you have any idea how to support that kind of tests after the change?
This

Reply:
Yes，In TiDB test a new test struct will override `ScanLock` method to generate expected locks.